src/main/java/net/schmizz/sshj/common/CircularBuffer.java

```diff
@@ -0,0 +1,151 @@
+            super(message);
+        }
+    }
+
+    public static final class PlainCircularBuffer
+            extends CircularBuffer<PlainCircularBuffer> {
+
+        public PlainCircularBuffer(int size, int maxSize) {
+            super(size, maxSize);
+        }
+    }
+
+    private int getNextSize(int i) {
+        // Use next power of 2.
+        int j = 1;
+        while (j < i) {
+            j <<= 1;
+            if (j <= 0) throw new IllegalArgumentException("Cannot get next power of 2; "+i+" is too large");
+        }
+        return Math.min(j, maxSize); // limit to max size
+    }
+
+    private final int maxSize;
+    private byte[] data;
+    private int rpos;
+    private int wpos;
+
+    public CircularBuffer(int size, int maxSize) {
+        this.maxSize = maxSize;
+        int initialSize = getNextSize(size);
+        if (initialSize > maxSize) {
+            throw new IllegalArgumentException(
+                String.format("Initial power of 2 size %d larger than maximum size %d", initialSize, maxSize));
+        }
+        this.data = new byte[initialSize];
+        this.rpos = 0;
+        this.wpos = 0;
+    }
+
+    public int available() {
+        int available = wpos - rpos;
+        return available >= 0 ? available : available + data.length; // adjust if wpos is left of rpos
+    }
+
+    public int maxPossibleRemainingCapacity() {
+        // Get remaining in current buffer.
+        int remaining = rpos - wpos;
+        if (remaining < 0) {
+            remaining += data.length; // adjust if rpos is left of wpos
+        }
+        // Add the maximum amount it can grow.
+        return remaining + maxSize - data.length;
+    }
+
+    private void ensureAvailable(int a)
+            throws CircularBufferException {
+        if (available() < a) {
+            throw new CircularBufferException("Underflow");
+        }
+    }
+
+    void ensureCapacity(int capacity) {
+        int available = available();
+        int remaining = data.length - available;
+        // If capacity fits exactly in remaining space, expand it; otherwise, wpos will equal rpos after writing.
+        if (remaining <= capacity) {
+            int cw = available + capacity + 1;
+            byte[] tmp = new byte[getNextSize(cw)];
+            // Copy the one or two segments between rpos and wpos and start again from 0.
+            if (wpos >= rpos) {
+                System.arraycopy(data, rpos, tmp, 0, available);
+                wpos -= rpos; // move wpos left so that rpos starts from 0
+            } else {
+                int tail = data.length - rpos;
+                System.arraycopy(data, rpos, tmp, 0, tail); // segment right of rpos
+                System.arraycopy(data, 0, tmp, tail, wpos); // segment left of wpos
+                wpos += tail; // move wpos right to let rpos wrap around to 0
+            }
+            rpos = 0;
+            data = tmp;
+        }
+    }
+
+    public void readRawBytes(byte[] buf, int off, int len)
```

Comment:
Block comment please. Is this reading from the circular buffer and placing the resulting data somewhere, or reading from somewhere and writing into the buffer? Given the name of `putRawBytes()`, I'm guessing it's the former.